src/common/src/array/list_array.rs

```diff
@@ -624,6 +626,230 @@ impl<'a> From<&'a ListValue> for ListRef<'a> {
+                        return Ok(None);
+                    }
+                    Ok(Some(self.parse_array()?.into()))
+                } else {
+                    self.parse_value()
+                }
+            }
+
+            /// Parse an array.
+            fn parse_array(&mut self) -> Result<ListValue, String> {
+                self.skip_whitespace();
+                if !self.try_consume('{') {
+                    return Err("Array value must start with \"{\"".to_string());
+                }
+                self.skip_whitespace();
+                if self.try_consume('}') {
+                    return Ok(ListValue::new(vec![]));
+                }
+                let mut elems = Vec::new();
+                loop {
+                    let mut parser = Self {
+                        input: self.input,
+                        data_type: self.data_type.as_list(),
+                    };
+                    elems.push(parser.parse()?);
+                    self.input = parser.input;
+
+                    // expect ',' or '}'
+                    self.skip_whitespace();
+                    match self.peek() {
+                        Some(',') => {
+                            self.try_consume(',');
+                        }
+                        Some('}') => {
+                            self.try_consume('}');
+                            break;
+                        }
+                        None => return Err(Self::eoi()),
+                        _ => return Err("Unexpected array element.".to_string()),
+                    }
+                }
+                Ok(ListValue::new(elems))
+            }
+
+            /// Parse a non-array value.
+            fn parse_value(&mut self) -> Result<Datum, String> {
+                if self.peek() == Some('"') {
+                    return Ok(Some(self.parse_quoted()?));
+                }
+                // peek until the next unescaped ',' or '}'
+                let mut chars = self.input.char_indices();
+                let mut has_escape = false;
+                let s = loop {
+                    match chars.next().ok_or_else(Self::eoi)? {
+                        (_, '\\') => {
+                            has_escape = true;
+                            chars.next().ok_or_else(Self::eoi)?;
+                        }
+                        (i, c @ ',' | c @ '}') => {
+                            let s = &self.input[..i];
+                            let trimmed = s.trim_end();
+                            // consume the value and leave the ',' or '}' for parent
+                            self.input = &self.input[i..];
+
+                            if trimmed.is_empty() {
+                                return Err(format!("Unexpected \"{c}\" character."));
+                            }
+                            if trimmed.eq_ignore_ascii_case("null") {
+                                return Ok(None);
+                            }
```

Comment:
Can we move all trim-related logic (incl `let trimmed =`) here in the `!has_escape` (aka `else`) branch?

Reply:
I meant to clarify the logic, rather than the code structure: trimming is only necessary when there is no escape.

Reply:
Sorry I misunderstood "the branch".